### PR TITLE
Separate waking the event loop, from communicating with Compositor

### PR DESCRIPTION
**Pull request by gterzian**
@paulrouget first step of #15934, Glutin only for now, please take a look...

If this makes sense, I will also update the code for ports other than Glutin...

One question: one do I add the `warn!` macro to `servolib`? Also perhaps we would also want to add `box`, since I had to switch to using `Box::new`...

---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17068)

### CODE REVIEW

**Comment by gterzian** [comment]
UPDATE: also added the implementation for the cef port...

File: ports/glutin/window.rs [hunk 1]
```diff
@@ -1047,17 +1046,27 @@ impl WindowMethods for Window {
...
         let window_proxy = create_window_proxy(self);
-
-        (box GlutinCompositorProxy {
-             sender: sender,
-             window_proxy: window_proxy,
-         } as Box<CompositorProxy + Send>,
-         box receiver as Box<CompositorReceiver>)
+        box GlutinEventLoopRiser {
+            window_proxy: window_proxy,
+        } as Box<EventLoopRiser + Send>
```
Comment (paulrouget) [inline ports/glutin/window.rs]: Is the `as …` necessary?

**Review by paulrouget — COMMENTED** [review]
Looks good to me. There are some nits to address. Also, I'm not 100% sure you need to use Box<CompositorProxy>.

Once these addressed, I'd like @asajeffrey to take a look as well.

File: ports/glutin/window.rs [hunk 2]
```diff
@@ -1047,17 +1046,27 @@ impl WindowMethods for Window {
...
+            fn rise(&self) {
+                // kick the OS event loop awake.
+                if let Some(ref window_proxy) = self.window_proxy {
+                    window_proxy.wakeup_event_loop()
+                }
+            }
+            fn clone(&self) -> Box<EventLoopRiser + Send> {
+                box GlutinEventLoopRiser {
+                    window_proxy: self.window_proxy.clone(),
+                } as Box<EventLoopRiser + Send>
```
Comment (paulrouget) [inline ports/glutin/window.rs]: Is the `as …` necessary?

**Comment by gterzian** [comment]
@paulrouget Ok, I replaced the traits with structs, and addressed the other comments too... 

I'm wondering if we need to add some sort of "clone" method to `CompositorReceiver`?

File: ports/cef/window.rs [hunk 3]
```diff
@@ -295,13 +295,18 @@ impl WindowMethods for Window {
...
 
-    fn create_compositor_channel(&self)
-                                 -> (Box<CompositorProxy+Send>, Box<CompositorReceiver>) {
-        let (sender, receiver) = channel();
-        (box CefCompositorProxy {
-             sender: sender,
-         } as Box<CompositorProxy+Send>,
-         box receiver as Box<CompositorReceiver>)
+    fn create_event_loop_riser(&self) -> Box<EventLoopRiser> {
+        struct CefEventLoopRiser
```
Comment (paulrouget) [inline ports/cef/window.rs]: Add a `;` at the end after `CefEventLoopRiser`.

File: ports/cef/window.rs [hunk 4]
```diff
@@ -295,13 +295,18 @@ impl WindowMethods for Window {
...
+        impl EventLoopRiser for CefEventLoopRiser {
+            fn rise(&self) {
+                app_wakeup();
+            }
+            fn clone(&self) -> Box<EventLoopRiser + Send> {
+                box CefEventLoopRiser {
+                } as Box<EventLoopRiser + Send>
```
Comment (paulrouget) [inline ports/cef/window.rs]: I think you can just do:
```rust
fn clone() {
  box CefEventLoopRiser
}
```

**Review by paulrouget — CHANGES_REQUESTED** [review]

**Comment by paulrouget** [comment]
Thank you Greg. This looks good to me now (beside a nit).

I'd like to get a review from Alan.

@asajeffrey - this is the first step to clarify the Embedder <-> Constellation communication. Can you take a look too?

**Review by paulrouget — APPROVED** [review]

File: components/compositing/compositor_thread.rs [hunk 5]
```diff
@@ -22,40 +22,54 @@ use style_traits::viewport::ViewportConstraints;
...
-/// to communicate with the compositor may have to kick OS event loops awake, communicate cross-
-/// process, and so forth.
-pub trait CompositorProxy : 'static + Send {
-    /// Sends a message to the compositor.
-    fn send(&self, msg: Msg);
-    /// Clones the compositor proxy.
-    fn clone_compositor_proxy(&self) -> Box<CompositorProxy + 'static + Send>;
+
+/// Used to wake up the event loop, provided by the servo port/embedder.
+pub trait EventLoopRiser : 'static + Send {
```
Comment (asajeffrey) [inline components/compositing/compositor_thread.rs]: Nit: not sure about the name. A riser is a part of a staircase :)

**Review by asajeffrey — APPROVED** [review]
LGTM, I just have an annoying bikeshedding name change. `EventLoopWakeup`?

File: components/compositing/compositor_thread.rs (hunk 5 above)
Comment (gterzian) [inline components/compositing/compositor_thread.rs]: How about `EventLoopAwaker`, so that we have a noun? https://en.wiktionary.org/wiki/awaker

**Comment by paulrouget** [comment]
```
ajeffrey> "Waker" would probably be more colloquial English: "Awaker" sounds like "I am more awake than you".
```

**Comment by gterzian** [comment]
Ok, `EventLoopWaker` it is...

**Comment by gterzian** [comment]
It looks like approving in Github isn't enough, and somebody needs to do a r+ at the bot...

**Comment by asajeffrey** [comment]
Hmm, I wonder why bors reckons this is still a WIP. Is it the unchecked boxes in the PR description?

**Comment by paulrouget** [comment]
Probably the title. I changed it. Do you need to re-approve?

**Comment by paulrouget** [comment]
https://github.com/bundler/homu/blob/288bda90d87f6500f7dd9dd779b633f7615f72e4/homu/main.py#L352

**Comment by gterzian** [comment]
@paulrouget @asajeffrey thanks for the review.

**Comment by paulrouget** [comment]
Thank you @gterzian

**Comment by asajeffrey** [comment]
@gterzian thanks!

**Pull request merged**